test/e2e/hooks_test.go

```diff
@@ -650,6 +650,122 @@ spec:
 		})
 }
 
+func (s *HooksSuite) TestTemplateLevelHooksWithRetry() {
+	var children []string
+	s.Given().
+		Workflow(`
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  name: retries-with-hooks-and-artifact
+  labels:
+    workflows.argoproj.io/test: "true"
+  annotations:
+    workflows.argoproj.io/description: |
+      when retries and hooks are both included, the workflow cannot resolve the artifact 
+    workflows.argoproj.io/version: '>= 3.5.0'
+spec:
+  entrypoint: main
+  templates:
+    - name: main
+      steps:
+        - - name: build
+            template: output-artifact
+            hooks:
+              started:
+                expression: steps["build"].status == "Running"
+                template: started
+              success:
+                expression: steps["build"].status == "Succeeded"
+                template: success
+              failed:
+                expression: steps["build"].status == "Failed" || steps["build"].status == "Error"
+                template: failed
+        - - name: print
+            template: print-artifact
+            arguments:
+              artifacts:
+                - name: message
+                  from: "{{steps.build.outputs.artifacts.result}}"
+    
+    - name: output-artifact
+      script:
+        image: python:alpine3.6
+        command: [ python ]
+        source: |
+          import time
+          import random
+          time.sleep(random.randint(0,20)) # lifecycle hook for running won't trigger unless it runs for more than "a few seconds"
+          with open("result.txt", "w") as f:
+            f.write("Welcome")
+          # sys.exit("fail!")
+      timeout: 10s
```

Comment:
With this construct

1. the test doesn't test the original issue (resolve artifact), the test will pass with/without the code change. We need to check if workflow phase succeed.
2. there's a chance all retries timeout and never execute `print-artifact`
3. it could add ~60s to the e2e test, is there a way to reduce this?